README.md
# TalkJs

Javascript interop for [TalkJs](https://talkjs.com)

See TalkJs [documentations](https://talkjs.com/docs/Getting_Started/index.html) documentations more infos

## Example

```dart
import 'dart:html';

import 'package:talkjs/talkjs.dart';

Future<void> main() async {
  addTalkJsScript();
  await TalkJs.ready;
  createInbox();
}

void createInbox() {
  final me = User(
    UserOptions()
      ..id = '12345'
      ..name = 'George Looney'
      ..email = 'george@looney.net'
      ..photoUrl = 'https://talkjs.com/docs/img/george.jpg'
      ..welcomeMessage = 'Hey there! How are you? :-)',
  );

  final talkSession = Session(
    SessionOptions()
      ..appId = '<APP_ID>'
      ..me = me,
  );

  final other = User(
    UserOptions()
      ..id = '54321'
      ..name = 'Ronald Raygun'
      ..email = 'ronald@teflon.com'
      ..photoUrl = 'https://talkjs.com/docs/img/ronald.jpg'
      ..welcomeMessage = 'Hey there! Love to chat :-)',
  );

  final conversationId = TalkJs.oneOnOneId(me, other);
  final conversation = talkSession.getOrCreateConversation(conversationId);
  conversation.setParticipant(me);
  conversation.setParticipant(other);

  final inbox = TalkJsInbox(
    talkSession.createInbox(
      InboxOptions()..selected = conversation,
    ),
  );

  final el = document.getElementById('talkjs-container');
  if (el == null) return;

  inbox.mount(el);
}
```

CHANGELOG.md
# Changelog

## 0.1.1

- Remove dart_browser_loader dependency

## 0.1.0

- Null safety
- Implement all TalkJS API

## 0.0.7
- fix off conversation

## 0.0.6

#### Inbox:
+ onConversationSelected stream when you select a conversation

## 0.0.5

+ Dart 2 stable

## 0.0.1

- Initial version.


lib/src/talk_js_base.dart
@JS()
library talk_js_base;

import 'dart:async';
import 'dart:html';
import 'dart:js' as js;

import 'package:js/js.dart';

import 'interop.dart' as interop;

const _talkjsScript = '''
(function(t,a,l,k,j,s){
s=a.createElement('script');s.async=1;s.src="https://cdn.talkjs.com/talk.js";a.head.appendChild(s)
;k=t.Promise;t.Talk={v:3,ready:{then:function(f){if(k)return new k(function(r,e){l.push([f,r,e])});l
.push([f])},catch:function(){return k&&new k()},c:l}};})(window,document,[]);
''';

external dynamic get undefined;

void addTalkJsScript() {
  js.context.callMethod('eval', [_talkjsScript]);
}

/// The main messaging UI.
/// Chats on the left, messages on the right.
/// Create an Inbox through [Session.createInbox] and then call [mount] to show it.
class TalkJsInbox {
  final interop.Inbox _talkJsInbox;

  late StreamController<interop.ConversationSelectedEvent>
      _onSelectConversationController;

  TalkJsInbox(this._talkJsInbox) {
    _onSelectConversationController = StreamController(
      onListen: _onListenSelectConversationController,
      onCancel: _onCancelSelectConversationController,
    );
  }

  Stream<interop.ConversationSelectedEvent> get onConversationSelected =>
      _onSelectConversationController.stream;

  /// Renders the UI inside a DOM element specified by [element].
  Future<void> mount(Element? element) {
    return promiseToFuture(_talkJsInbox.mount(element));
  }

  void on(String eventType, Function handler) {
    _talkJsInbox.on(eventType, allowInterop(handler));
  }

  void off(String eventType, Function handler) {
    _talkJsInbox.off(eventType, allowInterop(handler));
  }

  void destroy() {
    _talkJsInbox.destroy();
    _onSelectConversationController.close();
  }

  void selectById(String conversationId) {
    _talkJsInbox.select(conversationId);
  }

  void _onListenSelectConversationController() {
    on('conversationSelected', _onConversationSelected);
  }

  void _onCancelSelectConversationController() {
    // Commented because 'conversationSelected is an unknown event type
    // _talkJsInbox.off("conversationSelected", allowInterop((_) {}));
  }

  void _onConversationSelected(
    interop.ConversationSelectedEvent event,
  ) {
    _onSelectConversationController.add(event);
  }
}

class TalkJs {
  /// Complete when talkjs is ready
  static Future<void> get ready {
    return promiseToFuture(interop.ready);
  }

  ///
  /// NOTE: If this function changes make sure to update the documentation in `The_Talk_Object.md`
  ///
  /// A helper method to predictably compute a Conversation ID based on participants' ids in the given conversation.
  /// Use this method if you want to simply create a conversation between two users,
  /// not related to a particular product, order or transaction.
  ///
  /// The order of the parameters does not matter.
  /// For example, `Talk.oneOnOneId("a", "b")` yields the same result as `Talk.oneOnOneId("b", "a")`.
  ///
  /// This method takes the following steps:
  /// 1. Take two ids of users and put them in an array
  /// 2. Sort them lexicographically
  /// 3. JSON encode them
  /// 4. hash the result using SHA1, return the first 20 characters
  ///
  /// In pseudocode, this is what this function does:
  ///
  ///     var sorted = [me.id, other.id].sort()
  ///     var encoded = JSON.encode(sorted)
  ///     var hash = sha1(encoded)
  ///     return truncate(hash, 20)
  ///
  /// For a PHP implementation, see https://gist.github.com/eteeselink/4dc3ad32cc478986ff2b5b6361a1825f.
  /// {@link https://talkjs.com/?chat | Get in touch} if you need our help implementing this in your backend language.
  /// @public
  static String oneOnOneId(
    dynamic /*User|String*/ me,
    dynamic /*User|String*/ other,
  ) {
    return interop.oneOnOneId(me, other);
  }

  /// Retrieves global App metadata.
  ///
  /// @remarks This function may be called before instantiating a Talk.Session, and
  /// is not authenticated. Please consider all app metadata (eg global custom
  /// fields) to be public data, and ensure that you do not store any sensitive
  /// data in app custom fields.
  ///
  /// You can change App metadata using the TalkJS REST API.
  ///
  /// Note that this function cannot be used to verify whether an app ID exists; it
  /// always returns valid data, even for nonexistent app IDs.
  static Future<interop.AppMetadata> getAppMetadata(String appId) {
    return promiseToFuture(interop.getAppMetadata(appId));
  }
}


lib/src/talk_js_interop.dart
@JS()
library talk_js_interop;

import 'dart:html';
import 'package:js/js.dart';

@JS()
@anonymous
class UserOptions {
  external factory UserOptions({
    String id,
    String name,
    String? email,
    String? photoUrl,
    String? welcomeMessage,
    String? role,
  });
}

@JS('Talk.User')
class User {
  external factory User(UserOptions options);

  external String get id;

  external String get name;

  external String? get email;

  external String? get photoUrl;

  external String? get welcomeMessage;

  external String? role;
}

@JS('Talk.Inbox')
class InboxInterop {
  external void mount(Element element);

  external void destroy();

  external void on(String eventType, Function handler);

  external void off(String eventType, Function handler);

  external void select(
    dynamic /* ConversationBuilder|String*/ conversation, [
    InboxOptions? options,
  ]);
}

@JS('Talk.Chatbox')
class Chatbox {
  external void mount(Element element);

  external void destroy();
}

@JS()
@anonymous
class SessionOptions {
  external factory SessionOptions({
    String appId,
    User me,
    String? signature,
  });
}

@JS()
@anonymous
class ConversationAttributes {
  external factory ConversationAttributes({
    dynamic? custom,
    String? photoUrl,
    String? subject,
    List<String>? welcomeMessages,
  });
}

@JS()
@anonymous
class ParticipantSettings {
  external factory ParticipantSettings({
    String? access,
    bool? notify,
  });
}

@JS('Talk.ConversationBuilder')
class ConversationBuilder {
  external dynamic? get custom;
  external String? get photoUrl;
  external String? get subject;
  external List<String>? get welcomeMessages;

  external void setAttributes(ConversationAttributes attributes);
  external void setParticipant(
    User user,
    ParticipantSettings settings,
  );
}

@JS('Talk.Session')
class Session {
  external factory Session(SessionOptions options);

  external String get appId;

  external User get me;

  external ConversationBuilder getOrCreateConversation(String id);

  external InboxInterop createInbox(InboxOptions inboxOptions);

  external Chatbox createChatbox(ConversationBuilder conversation);

  external void syncThemeForLocalDev(String path);
}

@JS()
@anonymous
class InboxOptions {
  external dynamic /*ConversationBuilder|String*/ get selected;

  external set selected(dynamic /*ConversationBuilder|String*/ selected);

  external factory InboxOptions({
    dynamic /*ConversationBuilder|String*/ selected,
  });
}

@JS()
@anonymous
class ConversationData {
  external String get id;

  external String? get photoUrl;

  external String? get subject;

  external List<String>? get welcomeMessages;

  external factory ConversationData({
    String id,
    String? photoUrl,
    String? subject,
    List<String>? welcomeMessages,
  });
}

@JS()
@anonymous
class ConversationSelectedEvent {
  external ConversationData? get conversation;
  external User get me;
  external List<User>? get others;

  external factory ConversationSelectedEvent({
    ConversationData? conversation,
    User me,
    List<User>? others,
  });
}


example/main.dart
import 'dart:html';

import 'package:talkjs/talkjs.dart';

Future<void> main() async {
  addTalkJsScript();
  await TalkJs.ready;
  createInbox();
}

void createInbox() {
  final me = User(
    UserOptions()
      ..id = '12345'
      ..name = 'George Looney'
      ..email = 'george@looney.net'
      ..photoUrl = 'https://talkjs.com/docs/img/george.jpg'
      ..welcomeMessage = 'Hey there! How are you? :-)',
  );

  final talkSession = Session(
    SessionOptions()
      ..appId = '<APP_ID>'
      ..me = me,
  );

  final other = User(
    UserOptions()
      ..id = '54321'
      ..name = 'Ronald Raygun'
      ..email = 'ronald@teflon.com'
      ..photoUrl = 'https://talkjs.com/docs/img/ronald.jpg'
      ..welcomeMessage = 'Hey there! Love to chat :-)',
  );

  final conversationId = TalkJs.oneOnOneId(me, other);
  final conversation = talkSession.getOrCreateConversation(conversationId);
  conversation.setParticipant(me);
  conversation.setParticipant(other);

  final inbox = TalkJsInbox(
    talkSession.createInbox(
      InboxOptions()..selected = conversation,
    ),
  );

  final el = document.getElementById('talkjs-container');
  if (el == null) return;

  inbox.mount(el);
}
